### PR TITLE
[react-redux] fixed mistreating of action thunk creator parameters

### PR DESCRIPTION
**Pull request by surgeboris**
PR addresses https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30740

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30740>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/react-redux/index.d.ts
Comment (Kovensky) [inline types/react-redux/index.d.ts]: Interesting, I thought argument type tuples were a 3.1 feature.

Looks like the part that _is_ a 3.1 feature would be mapping on the tuple, but luckily we don't need to map it.

**Review by Jessidhia — APPROVED** [review]

**Comment by surgeboris** [comment]
Need adjustments for the usecase described in https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30816

Will implement them once I'll have time.

**Comment by surgeboris** [comment]
PR has been adjusted according to usecase described in #30816.

I think it's ready to merge, but if someone wants to review it - I'll be very glad if you do so.

I'm also opened to any suggestions on how we can re-organize overloads, and make type definitions more readable.

File: types/react-redux/react-redux-tests.tsx
Comment (Kovensky) [inline types/react-redux/react-redux-tests.tsx]: If you're using `redux-thunk` you should be using `ThunkDispatch` as the type of `dispatch` which would correctly handle thunk action creators.

Probably better to add a test just to make sure `mapDispatchToProps(dispatch: ThunkDispatch)` is assignable as the 2nd argument.

File: types/react-redux/react-redux-tests.tsx
Comment (Kovensky) [inline types/react-redux/react-redux-tests.tsx]: It could be nice to have a simply-named type operator that you can use to transform the types yourself.

This is how I generally declare my components' inputs:

```ts
interface OwnProps {
  // ...
}

interface ReduxProps {
  // things from mapStateToProps
}

function mapStateToProps(state: State /* if required: , ownProps: OwnProps */): ReduxProps {
  return { /* ... */ }
}

const mapDispatchToProps = {
  actionCreator1,
  actionCreator2
}

type Props = OwnProps & ReduxProps & typeof mapDispatchToProps
```

The request is for something like `type ResolveThunks<T> = ...` that you could then use like `ResolveThunks<typeof mapDispatchToProps>`.

Comment (Kovensky) [reply types/react-redux/react-redux-tests.tsx]: (oh, you have `WithThunkActionCreators` but would be nice to have a shorter name)

**Review by Jessidhia — APPROVED** [review]

File: types/react-redux/react-redux-tests.tsx
Comment (surgeboris) [reply types/react-redux/react-redux-tests.tsx]: Can I use a `ThunkDispatch` type in tests without adding explicit dependency on `redux-thunk`?

If "No", then what are the benefits  of using `ThunkDispatch` instead of simple `Dispatch` in the test? As far as I can tell `react-redux` will NOT use type inference to pick up `ThunkDispatch` type from `mapDispatchToProps` argument (i.e. `react-redux` will still use simple `Dispatch` under the hood  regardless of what type you're using in test).

I'd like to avoid direct explicit dependency on `redux-thunk`, because:
- `react-redux` can easily be used without `redux-thunk`, so it's there's no actual depenendency in place (and I'm not sure if adding `redux-thunk` to `react-redux`'s  `package.json` will not lead to problem for the people not using `redux-thunk`)
- even if we use `redux-thunks` only for the tests, future maintainers may use it in the type definitions (just because the dependency is already there), which may make them less generic (they will stop working for custom `redux-thunk` implemenations, for example)
- it will be a bad example for the future maintainers that may add dependencies on other Redux middlewares by analogy which may affect `redux-thunk`'s definitions verstaility/maintainability (I cannot imagine how though; it more like a gut feeling).

File: types/react-redux/react-redux-tests.tsx
Comment (surgeboris) [reply types/react-redux/react-redux-tests.tsx]: Renamed `WithActionCreators` to `ResolveThunks`.

File: types/react-redux/react-redux-tests.tsx
Comment (Kovensky) [reply types/react-redux/react-redux-tests.tsx]: Not really...... would probably have to copy the definition

```ts
export interface ThunkDispatch<S, E, A extends Action> {
  <T extends A>(action: T): T;
  <R>(asyncAction: ThunkAction<R, S, E, A>): R;
}
```

Because it doesn't `extends Dispatch` it looks like it wouldn't work, not directly. With the way the `Middleware` / `StoreEnhancer` interface works, they actually can upgrade the type of `dispatch` in a specific `redux` store (which would have to be accessed with `typeof store['dispatch']`), but I'm not sure there's any way to propagate it to `react-redux` as `connect` doesn't know what store it is connecting to.

So there are two alternatives: do something similar to what `styled-components` does and have a way to reexport in your own code a version of `react-redux` where the store/dispatch types are hardcoded to types you're aware of, or change `redux-thunk` to directly augment `Dispatch<A>` instead.

Or recommend that users write the module augmentation themselves:

```ts
import { ThunkDispatch } from 'redux-thunk'

interface State {
  // your redux store state
}

declare module 'redux' {
  interface Dispatch<A extends Action> extends ThunkDispatch<State, never, A> {}
}
```

Comment (surgeboris) [reply types/react-redux/react-redux-tests.tsx]: Thank you very much for explanation, but to be honest I did not get the complete picture out of it (seems like I'm not yet ready for such an abstract types; I'll probably need to re-read your comment again and again to understand it).

Even if the problem is real (I'll check if it exists for me in my project), it seems like it was there even before I introduced any of my thunk-changes - hence it's not immediate subject of this PR. I'm going to contemplate on it and maybe I'll come up with the follow up PR to address it.

Right now I'm more concerned by the fact that my previous PR introduced regression into the @types/react-reduc@6.0.10, so I'd like to merge this PR as fast as possible to fix previously introduced regression.

So, if you don't mind, let's delay this problem for a while.

**Review by Jessidhia — APPROVED** [review]
Would be nice to get this merged ASAP 🔥

**Comment by Kovensky** [comment]
There is a problem when whatever is given to `mapDispatchToProps` itself has multiple signatures or generic arguments; those are all erased by the mapping. This is a limitation of TypeScript, not sure there's anything we can do about it.

**Comment by Kovensky** [comment]
Yeah, I still think it's ok, it's just something I noticed when I tried making a function that would do the same kind of parameter transformation.

File: types/mirrorx/index.d.ts
Comment (amcasey) [inline types/mirrorx/index.d.ts]: Note that increasing the TS version prevents consumers on older versions from using these type definitions.

Comment (Kovensky) [reply types/mirrorx/index.d.ts]: Yes, and this will be the case up to TS 3.0.

It has to be bumped because react-redux itself was bumped, and http://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#generic-rest-parameters is needed for this to work correctly.

**Pull request merged**
